build.gradle

```diff
@@ -13,187 +13,154 @@
 }
 
 subprojects {
-  apply plugin: 'java'
-  apply plugin: 'maven'
-  apply plugin: 'signing'
-
-  group = 'com.cloudant'
-  version = new File(rootDir, 'VERSION').text.trim()
-
-  // If the version says "snapshot" anywhere assume it is not a release
-  ext.isReleaseVersion = !version.toUpperCase(Locale.ENGLISH).contains("SNAPSHOT")
-
-  // 1.8 is ok for boot 1.x and 2.x
-  sourceCompatibility = 1.8
-  targetCompatibility = 1.8
-
-  repositories {
-    mavenLocal()
-    mavenCentral()
-  }
-
-  // Common dependencies
-  ext.compileWithSpringVersion = '5.0.6.RELEASE'
-  ext.compileWithBootVersion = '2.0.2.RELEASE'
-
-  ext.testWithSpringVersion = '4.3.17.RELEASE'
-  ext.testWithBootVersion = '1.5.13.RELEASE'
-  
-  dependencies {
-    // java-cloudant client
-    compile group: 'com.cloudant', name: 'cloudant-client', version: '2.19.1'
-    // spring
-    compile group: 'org.springframework', name: 'spring-core', version: compileWithSpringVersion
-    compile group: 'org.springframework', name: 'spring-context', version: compileWithSpringVersion
-    compile group: 'org.springframework', name: 'spring-beans', version: compileWithSpringVersion
-      // testing
-    testCompile group: 'junit', name: 'junit', version: '4.12'
-    testCompile group: 'org.springframework.boot', name: 'spring-boot-starter-test', version: testWithBootVersion    
-    }
+    apply plugin: 'java-library'
+    apply plugin: 'maven-publish'
+    apply plugin: 'signing'
+    apply plugin: 'com.github.spotbugs'
 
-  // Include variable debug info in the compiled classes
-  compileJava.options.debugOptions.debugLevel = "source,lines,vars"
-  // Fail on javac warnings
-  compileJava.options.compilerArgs << "-Werror"
-
-  // Always UTF-8
-  tasks.withType(JavaCompile) {
-    options.encoding = 'UTF-8'
-  }
-
-  task sourcesJar(type: Jar, dependsOn: classes) {
-    classifier = 'sources'
-    sourceSets.all {
-      into(name + "/java", { from allJava })
-      into(name + "/resources", { from resources })
-    }
-  }
-
-  javadoc {
-    options.encoding 'UTF-8'
-    options.setMemberLevel JavadocMemberLevel.PROTECTED
-    // Add a logging listener to check for javadoc warnings and fail the build if there are any
-    boolean hasJavaDocWarnings = false;
-    doFirst {
-      getLogging().addStandardErrorListener(new StandardOutputListener() {
-        void onOutput(CharSequence output) {
-          if (output =~ "warning:") {
-            hasJavaDocWarnings = true
-          }
-        }
-      })
-    }
-    doLast {
-      if (hasJavaDocWarnings) {
-        throw new GradleException("Build failed due to javadoc warnings.");
-      }
+    group = 'com.cloudant'
+    version = new File(rootDir, 'VERSION').text.trim()
+
+    // If the version says "snapshot" anywhere assume it is not a release
+    ext.isReleaseVersion = !version.toUpperCase(Locale.ENGLISH).contains("SNAPSHOT")
+
+    sourceCompatibility = 11
+    targetCompatibility = 17
```

Comment:
Any particular reason to target 17 here? I agree that's fine from our test perspective, but in terms of binaries we ship I feel like it ought to be `11` (or even `8`)

Reply:
> As of Spring Framework 6.0, Spring requires Java 17+.

(from https://docs.spring.io/spring-framework/docs/current/reference/html/overview.html)

So I guess we just need to have a discussion about minimum spring/java versions we support.

Reply:
I'm going to suggest Spring Framework 5.3.27 and Spring Boot 2.7.9

Framework [5.3.x is documented](https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Versions#supported-versions) as:
>5.3.x is the final feature branch of the 5th generation, with long-term support provided on JDK 8, JDK 11, JDK 17 and the Java EE 8 level.

Boot [2.7.9 claims](https://docs.spring.io/spring-boot/docs/2.7.9/reference/htmlsingle/#getting-started.system-requirements):
>Spring Boot 2.7.9 requires Java 8 and is compatible up to and including Java 19. [Spring Framework 5.3.25](https://docs.spring.io/spring-framework/docs/5.3.25/reference/html/) or above is also required.

which I think should allow us to use Java 17, but with 1.8 source and target compatibility. I think that will give us slightly less churn and give us the best chance of utilizing the existing work from #22 and #23.